### ISSUE TITLE
Gradient of nansum and nanmean wrongly produces `nan`

### ISSUE THREAD

**Issue by mfkasim1** [pytorch/pytorch#67180]
## 🐛 Bug

Tensors that are multiplied with another tensor containing `nan`, then going through `nansum` or `nanmean` operator returns `nan` as its gradient.
To be clear, just take a look at the code below.

## To Reproduce

```python
import torch
a = torch.tensor([1., 2., float('nan')])
b = torch.tensor(1.0, requires_grad=True)
c = a * b
c1 = torch.nansum(c)  # or torch.nanmean

bgrad1, = torch.autograd.grad(c1, b, retain_graph=True)
print(bgrad1)
```
which prints
```
tensor([nan])
```

## Expected behavior

The variable `c1` can be expressed as `c1 = b * a[0] + b * a[1]` (no `a[2]` because it's `nan`), so the derivative of `c1` w.r.t. `b` is simply `a[0] + a[1]` which should not be nan.

## Environment

 - PyTorch Version (e.g., 1.0): 1.10.0+cpu
 - OS (e.g., Linux): Ubuntu 20.04 WSL
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): N/A
 - Python version: 3.8.5
 - CUDA/cuDNN version: N/A
 - GPU models and configuration: N/A
 - Any other relevant information: -

## Additional context

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by george-qi** [comment]
Hi all, adding on to what @cpuhrsch mentioned above: wanted to surface some attention for MaskedTensor, a tensor extension library that we just launched the prototype for along PyTorch 1.11. It should help solve many existing problems related to masked semantics, distinguishing between 0 and undefined gradients, etc, and indeed, it should fix this issue as per the above code snippet.

For more details, you can check out the [website](pytorch.org/maskedtensor) or [Github](github.com/pytorch/maskedtensor). We're actively looking for new users, so if you're interested in using it or have any suggestions on things we should build out, please reach out to me or @cpuhrsch. Thanks!!

**Comment by soulitzer** [comment]
Duplicate of https://github.com/pytorch/pytorch/issues/4132
